### DISCUSSION TITLE
Active Copilot suggestions disable IdeaVim mode switching

### DISCUSSION THREAD

**Discussion by danielfrankcom** [community/community#7863] [Copilot Conversations] [answered]
I am using Copilot in IntelliJ, alongside the IdeaVim plugin.

When a Copilot suggestion is visible (greyed out ghost code before the suggestion is accepted with \<tab>), the \<esc> key rejects the suggestion. When using IdeaVim, one must frequently switch from insert mode to normal mode using the <esc> key. Since suggestions are often made when a line is finished being typed, this means that the \<esc> key needs to be pressed twice to enter normal mode, which breaks with previous muscle memory of using vim.

This has resulted in a number of situations where I try to move around or change a word after hitting \<esc>, and find that the characters I'm pressing are being typed because I'm still in insert mode.

My suggestion would be to have the first \<esc> apply to the IdeaVim plugin, so that code editing using vim shortcuts is not interrupted by Copilot suggestions. Currently, if a suggestion is visible and one uses the arrow keys (still in insert mode) then the suggestion disappears. This functionality could be maintained in normal mode, where the first \<esc> press moves from insert mode to normal mode, and any future key pressed (j/k to move down/up) would indicate that the Copilot suggestion should disappear, just like when the arrow keys are pressed for non-vim code navigation.

IdeaVim version: 1.8.1
Copilot version: 1.0.3
IntelliJ version: 2021.2.3

**Comment by kurtbuilds** [comment, marked as answer]
This was bothering me too. You can work around it quite easily by modifying the keyboard shortcut for Github Copilot -> Hide Completions in Editor from the default value of escape. I updated it to cmd+escape, so it's still easy to dismiss the suggestions, but it no longer interferes with vim mode escape to normal mode.

<img width="909" alt="Screen Shot 2021-11-19 at 13 32 50" src="https://user-images.githubusercontent.com/79804461/142673803-671f161d-7219-4059-9b9c-280bead2cfb6.png">

**Reply by danielfrankcom** [reply]
Ah perfect, this works great. It would be nice if this integration worked out of the box, but this is certainly a functional workaround. Thanks!

**Reply by kurtbuilds** [reply]
Agreed. IntelliJ is good about warning users when keymaps conflict, but it doesn't work because modal editing functions doesn't use the IntelliJ hotkey system.

This bug definitely harms the first-time UX for Copilot, so I think they should have enable a conflict warning in the Keymap editor specifically for the escape key if the user has IdeaVim enabled, and/or a warning about it when you install Copilot.

> Ah perfect, this works great. It would be nice if this integration worked out of the box, but this is certainly a functional workaround. Thanks!

**Reply by jansorg** [reply]
@danielfrankcom Thanks for reporting! So far, there's no integration with IdeaVim mode, but it certainly shouldn't break it. Could you update the shortcuts of the 1st paragraph? Markdown seems to have swallowed most of them and it's hard to tell which keys you're referring to. 
We're tracking this internally now if it's possible to change the handling of ESC with IdeaVim mode.

**Reply by danielfrankcom** [reply]
I just fixed the shortcuts in the original post, thanks for pointing that out.

As a follow-up after using this for a couple of days, the behavior is still a bit weird even after remapping the "Hide Completions in Editor" shortcut. When a suggestion shows up in insert mode and you back out to normal mode everything is fine because the act of doing that clears the suggestion. There are some situations however where a suggestion will show up while you are in normal mode (after deleting a line for example using \<dd>), and then you cannot use certain Vim shortcuts. Specifically, the 4 up/down/left/right keys have no effect, locking you in place. Interestingly, other keys do work, such as the \<w> and \<b> key to move forward/back 1 word.

Having cleared the "Hide Completions in Editor" shortcut so that it is not bound, I am running into problems less, but I still do run into issues as described above when a suggestion shows up in normal mode.

**Reply by jansorg** [reply]
Thanks for your follow-up! We'll have to investigate the use with IdeaVim a bit to improve the keyboard handling.

**Reply by kurtbuilds** [reply]
I’m also infrequently running into the same issue @danielfrankcom describes. Copilot will fill suggestions even when not in insert mode, which I think is wrong behavior.

**Reply by danielfrankcom** [reply]
Personally I don't mind that it suggests things in normal mode, I just don't want it to prohibit my Vim intentions at any given point. I'd like to be able to _accept_ a suggestion at any point by pressing \<tab>, but otherwise I want it to stay out of my way and let me move around and do what I'd otherwise like to.

**Reply by jansorg** [reply]
The next update will contain a fix to not override IdeaVim shortcuts. I'm not a IdeaVim user myself, please let us know if there are any other issues breaking your IdeaVim flow.

**Reply by danielfrankcom** [reply]
I just pulled the latest update (1.1.3.139) and reset my keymap to test the stock experience. The issue with \<esc> appears to be fixed, and a single press of \<esc> will now move from insert mode to normal mode, which is perfect. Thanks for fixing that conflict!

Unfortunately, the other key conflict that I mentioned still exists in this version. Sometimes when you delete a line using \<dd> you can trigger Copilot to make a code suggestion while you are in normal mode, which is fine by itself. The problem is that when this suggestion is active in normal mode, you cannot use the navigation keys (\<j>, \<k>, \<h>, \<l>) to move around until you manually press \<esc> to clear the suggestion. For some reason it seems that other Vim shortcuts still work in this scenario, such as \<w> and \<b>. I noticed that I also cannot move using the arrow keys, which are pretty much the same thing as the 4 navigation keys in normal mode. 

What I would expect to happen in this scenario is what happens in insert mode when you use the arrow keys. If a suggestion is visible and you press the arrow keys, the cursor moves and the suggestion disappears. I cannot test the Vim navigation keys in insert mode because those keys do something different in that mode, but there is a distinct difference in the behavior of the arrow keys in these 2 modes. In insert mode, the arrow keys do what I would expect, and clear the suggestion as the cursor moves. In normal mode, the arrow keys do nothing, and the cursor is locked in place until the suggestion is manually cleared with \<esc>.

The latest update is certainly an improvement for this integration, so thanks again, but there are still some thorns in certain scenarios as mentioned. The behavior of having a Copilot suggestion block movement around the file is quite intrusive, as it breaks your train of thought and adds an additional manual step that needs to be performed to move the cursor. I haven't been able to find a way to work around this issue yet, so I'm still running into it when modifying code.

**Reply by jansorg** [reply]
@danielfrankcom Thank you for your feedback! It definitely shouldn't block other shortcuts. We'll investigate how to fix that.

**Reply by jansorg** [reply]
@danielfrankcom Could you add which OS, IDE, and IDE version you're currently using?
I failed to reproduce this (Linux, 2021.2 + IdeaVIM 0.7.2 / 0.9.0, 2021.3 + IdeaVIM 0.9.0). I was able to move the caret using <j>, <k>, <h>, <l> and also with the arrow keys.
If you request Copilot suggestions manually (`Alt+\`), are keys blocked in the same way?

Are Copilot suggestions in normal mode useful in any way? Unless you change the shortcut to insert suggestions into the editor you wouldn't be able to use them (TAB as default shortcut doesn't seem to work in normal mode).
If it's not that useful, I'd just turn off Copilot suggestions in normal mode. Would that fix your problems with the plugin?

Thanks!

**Reply by danielfrankcom** [reply]
I'm on Windows 10 (19042.1348), using IntelliJ IDEA 2021.3 (Ultimate Edition, Build #IU-213.5744.223), IdeaVim 1.8.1, Copilot 1.1.3.139. I was also able to reproduce this issue on IntelliJ 2021.2.3 before I upgraded, using the same versions of everything else. IdeaVim just pushed an update today so I upgraded to 1.9.1, however I'm still seeing the issue with that version too.

There are some shortcuts missing from your message, I think for the same markdown reasons as my original message. I had to use a backslash to escape the `<` character when typing shortcuts in my messages.

I notice that the IdeaVim version that you are using doesn't really line up with the version numbers that I'm seeing, so I wonder if there is a possible issue there? The [IdeaVim plugin page](https://plugins.jetbrains.com/plugin/164-ideavim/versions) seems to indicate that I'm using the newest version.

I tried manually requesting suggestions using `Alt+\` (while in normal mode) and to my surprise I was able to move around, meaning that everything worked as expected. The keys only seem to be blocked when the suggestions come up organically. It's possible that this only shows up in a subset of normal mode conditions, but the only scenario where I've been able to get a suggestion to come up organically is when I delete a line, so I haven't tested any other situations.

My previous comment about not minding suggestions in normal mode (providing the keys aren't blocked) still holds true, they wouldn't harm anything in theory, but ask you ask, I don't see a lot of value for them. I actually didn't realize that you couldn't use \<TAB> to autocomplete in normal mode because I've never seen a suggestion useful enough to accept after deleting a line. I don't want to speak for anyone else, but I wouldn't be disappointed if suggestions were disabled in normal mode completely. I think that would fix my problem because the suggestions would never show up, and would therefore never block movement.

**Reply by jansorg** [reply]
@danielfrankcom Thank you for your feedback! It's odd that manually requested suggestions don't block - there shouldn't be a difference. IdeaVIM is hooking deep into the shortcut handling and GitHub Copilot is similar, which may cause incompatibilities. 
For the next update we've merged a change to automatically turn off Copilot suggestions when you're in normal mode. If that still isn't fixing things properly, please let us know!

**Reply by jansorg** [reply]
The new update is available now, please let us know if it's not working for you.

**Reply by danielfrankcom** [reply]
The latest update seems to work perfectly, I can't get it to suggest anything in normal mode. I tried to get suggestions to show up manually (using `Alt+\`), and also automatically by deleting lines using \<dd>. Suggestions that show up in insert mode go away when pressing \<ESC> to transition to normal mode.

Really appreciate the fast feedback/response cycle on this!

**Reply by wolph** [reply]
It still gets stuck at times. I can't reliably reproduce it so I'm guessing it's a race condition, but at times it's completely stuck in insert mode and I have to disable/re-enable ideavim to get back to normal mode again.

**Reply by jansorg** [reply]
@WoLpH Could you add which version of GitHub Copilot you're using? Are you sure that this only happens when the GitHub Copilot plugin is enabled? 
When this happens again, could you please attach the log opened by `Help > Show log in ...`?

**Reply by wolph** [reply]
I'm using "Github Copilot 1.1.4.220" in "PyCharm 2021.3.1 / Build #PY-213.6461.77, built on December 28, 2021".
Unfortunately the issue is not easily reproducible so it's hard to say if it only happens when the Github Copilot plugin is enabled, but I will disable it for several days and see if it still occurs.

I should note that I had previously already removed the Esc shortcut before the update and was already noticing these issues with the older version so it might be a distinct issue. The Esc shortcut is currently enabled for me:
![image](https://user-images.githubusercontent.com/270571/150185496-89725448-93d5-4016-93ba-e9523b8357d5.png)

In any case, I'll try without the plugin for several days. It's always possible that PyCharm is doing something weird of course. Wouldn't be the first time :)

**Reply by wolph** [reply]
Perhaps it's related to this. Even after pressing the <kbd>Esc</kbd> button the completions are still visible at times. Note the cursor at the line below while it still shows the completion. Repeatedly pressing <kbd>Esc</kbd> makes no difference.
![image](https://user-images.githubusercontent.com/270571/150580539-38f4465a-c14c-4a88-8e98-7b61d3d2d734.png)

Below are the logs for the last 10 minutes in which I have repeatedly reproduced this:
```
2022-01-21 19:20:57,252 [56770294]   WARN - a.repo.GitUntrackedFilesHolder - Ignoring ignored files under another root: 16 files total 
2022-01-21 19:21:02,587 [56775629]   WARN - efaultCopilotCompletionService - [2794] Response received. Duration: 331 ms 
2022-01-21 19:21:19,959 [56793001]   INFO - me.idea.vim.ui.ex.ExEntryPanel - Activate ex entry panel 
2022-01-21 19:21:20,601 [56793643]   INFO - me.idea.vim.ui.ex.ExEntryPanel - Deactivate ex entry panel 
2022-01-21 19:21:37,307 [56810349]   WARN - efaultCopilotCompletionService - [2828] Response received. Duration: 2003 ms 
2022-01-21 19:21:37,400 [56810442]   WARN - efaultCopilotCompletionService - [2828] Response received. Duration: 2097 ms 
2022-01-21 19:21:37,443 [56810485]   WARN - efaultCopilotCompletionService - [2828] Response received. Duration: 2140 ms 
2022-01-21 19:21:53,633 [56826675]   WARN - efaultCopilotCompletionService - [2835] Response received. Duration: 1009 ms 
2022-01-21 19:21:53,637 [56826679]   WARN - efaultCopilotCompletionService - [2835] Response received. Duration: 1013 ms 
2022-01-21 19:21:53,727 [56826769]   WARN - efaultCopilotCompletionService - [2835] Response received. Duration: 1103 ms 
2022-01-21 19:22:08,711 [56841753]   WARN - efaultCopilotCompletionService - [2858] Response received. Duration: 208 ms 
2022-01-21 19:22:15,330 [56848372]   WARN - efaultCopilotCompletionService - [2860] Response received. Duration: 2170 ms 
2022-01-21 19:22:15,423 [56848465]   WARN - efaultCopilotCompletionService - [2860] Response received. Duration: 2263 ms 
2022-01-21 19:22:15,596 [56848638]   WARN - efaultCopilotCompletionService - [2860] Response received. Duration: 2436 ms 
2022-01-21 19:22:21,380 [56854422]   INFO - er.VisualOperatorActionHandler - Execute visual command Command(rawCount=0, action=com.maddyhome.idea.vim.action.change.delete.DeleteVisualAction@34732f5, type=DELETE, flags=[FLAG_EXIT_VISUAL]) 
2022-01-21 19:23:05,811 [56898853]   WARN - efaultCopilotCompletionService - [2889] Response received. Duration: 417 ms 
2022-01-21 19:23:07,088 [56900130]   WARN - efaultCopilotCompletionService - [2893] Response received. Duration: 182 ms 
2022-01-21 19:23:15,328 [56908370]   WARN - efaultCopilotCompletionService - [2901] Response received. Duration: 210 ms 
2022-01-21 19:23:27,019 [56920061]   WARN - ConfigurableExtensionPointUtil - ignore configurable with duplicated id: Jupyter Settings 
2022-01-21 19:23:27,206 [56920248]   INFO - eStyle.CustomCodeStyleSettings - Loaded com.intellij.psi.formatter.xml.XmlCodeStyleSettings 
2022-01-21 19:24:11,058 [56964100]   WARN - tions.ex.ConfigurableCardPanel - auto-dispose 'Plugins' id=preferences.pluginManager 
2022-01-21 19:24:20,257 [56973299]   WARN - efaultCopilotCompletionService - [2907] Response received. Duration: 244 ms 
2022-01-21 19:24:41,460 [56994502]   WARN - efaultCopilotCompletionService - [2916] Response received. Duration: 206 ms 
2022-01-21 19:24:52,155 [57005197]   INFO - il.indexing.FileBasedIndexImpl - Health check heartbeat 
2022-01-21 19:29:03,411 [57256453]   INFO - j.ide.actions.RevealFileAction - Exit code 1 
```

Version info:
```
2022-01-21 03:34:47,697 [    739]   INFO - llij.ide.plugins.PluginManager - Loaded custom plugins: CSV (2.18.2), Dash (3.3), ANSI Highlighter (1.2.4), IdeaVim (1.9.3), .ignore (4.3.0), Live Edit (213.5744.122), Requirements (2022.1-2021.3), Mypy (0.12.1), GitHub Copilot (1.1.6.322), Kubernetes (213.6461.19), .env files support (2021.3.0.213), CMD Support (1.0.5), Batch Scripts Support (1.0.13) 
PyCharm 2021.3.1 (Professional Edition)
Build #PY-213.6461.77, built on December 28, 2021
```

I should note that I have seen this issue both on Windows and OS X.

**Reply by wolph** [reply]
It (not being able to exit insert mode) just happened to me again, and this is the log output around the occurence:
```
2022-01-22 01:57:18,870 [80551912]   INFO - er.VisualOperatorActionHandler - Execute visual command Command(rawCount=0, action=com.maddyhome.idea.vim.action.copy.YankVisualAction@5dd1c7db, type=COPY, flags=[FLAG_EXIT_VISUAL]) 
2022-01-22 01:57:36,752 [80569794]   INFO - er.VisualOperatorActionHandler - Execute visual command Command(rawCount=0, action=com.maddyhome.idea.vim.action.change.delete.DeleteVisualAction@34732f5, type=DELETE, flags=[FLAG_EXIT_VISUAL]) 
2022-01-22 01:57:45,903 [80578945]   WARN - efaultCopilotCompletionService - [4265] Response received. Duration: 445 ms 
2022-01-22 01:57:48,134 [80581176]   WARN - efaultCopilotCompletionService - [4266] Response received. Duration: 2136 ms 
2022-01-22 01:57:48,162 [80581204]   WARN - efaultCopilotCompletionService - [4266] Response received. Duration: 2165 ms 
2022-01-22 01:57:48,179 [80581221]   WARN - efaultCopilotCompletionService - [4266] Response received. Duration: 2182 ms 
2022-01-22 01:58:15,232 [80608274]   WARN - efaultCopilotCompletionService - [4293] Response received. Duration: 251 ms 
2022-01-22 01:58:23,980 [80617022]   WARN - efaultCopilotCompletionService - [4299] Response received. Duration: 180 ms 
2022-01-22 01:59:10,602 [80663644]   WARN - efaultCopilotCompletionService - [4301] Response received. Duration: 273 ms 
2022-01-22 01:59:11,277 [80664319]   WARN - efaultCopilotCompletionService - [4302] Response received. Duration: 357 ms 
2022-01-22 01:59:20,891 [80673933]   INFO - j.ide.actions.RevealFileAction - Exit code 1 
```

**Reply by wolph** [reply]
And this time it happened on the OS X machine and I think I might have gotten some useful logs:
```
2022-01-26 13:10:54,469 [77135343]   INFO - cloudConfig.CloudConfigManager - === Start.updateConnection === 
2022-01-26 13:10:54,469 [77135343]   INFO - cloudConfig.CloudConfigManager - === doConnect === 
2022-01-26 13:10:54,469 [77135343]   INFO - cloudConfig.CloudConfigManager - === getConnectionState === 
2022-01-26 13:14:41,311 [77362185]   INFO - il.indexing.FileBasedIndexImpl - Health check heartbeat 
2022-01-26 13:14:51,973 [77372847]   WARN - efaultCopilotCompletionService - fetchCompletions: unable to create file for BasicEditorRequest(language=com.github.copilot.lang.python.PythonLanguageSupport@49e6ec11, completionType=GhostText, useTabIndents=false, tabWidth=4, requestId=4431, fileLanguage=Language: Python, relativeFilePath=[... redacted ... ]/base.py, documentContent=import abc

[...random code snippet here...]

, offset=42, lineInfo=LineInfo(lineCount=15, lineNumber=4, lineStartOffset=38, columnOffset=4, line=    , nextLineIndent=4), requestTimestamp=1643199291639, documentModificationSequence=137, isCancelled=true) 
2022-01-26 13:14:54,181 [77375055]   WARN - efaultCopilotCompletionService - [4440] Response received. Duration: 261 ms 
2022-01-26 13:14:54,182 [77375056]   WARN - efaultCopilotCompletionService - [4440] Response received. Duration: 262 ms 
2022-01-26 13:14:54,183 [77375057]   WARN - efaultCopilotCompletionService - [4440] Response received. Duration: 263 ms 
2022-01-26 13:14:57,082 [77377956]   WARN - efaultCopilotCompletionService - fetchCompletions: unable to create file for BasicEditorRequest(language=com.github.copilot.lang.python.PythonLanguageSupport@49e6ec11, completionType=GhostText, useTabIndents=false, tabWidth=4, requestId=4449, fileLanguage=Language: Python, relativeFilePath=[... redacted ... ]/base.py, documentContent=import abc

[...random code snippet here...]

, offset=81, lineInfo=LineInfo(lineCount=17, lineNumber=6, lineStartOffset=67, columnOffset=14, line=        self.c, nextLineIndent=4), requestTimestamp=1643199297002, documentModificationSequence=156, isCancelled=true) 
2022-01-26 13:14:57,509 [77378383]   WARN - efaultCopilotCompletionService - [4452] Response received. Duration: 163 ms 
2022-01-26 13:14:59,180 [77380054]   WARN - efaultCopilotCompletionService - [4462] Response received. Duration: 238 ms 
```

**Reply by jansorg** [reply]
@WoLpH Thanks! I'll look into it. Are you unable to leave IdeaVIM's insert mode or does that work, but the Copilot  hints remain visible?
Which version of IdeaVIM are you using?

**Reply by wolph** [reply]
With the earlier cases the copilot hints were still visible, in this case there were no visible hints.

The main issue appears to be that the <btn>esc</btn> button stops working. 
I could leave insert mode by switching to another application, but when returning the <btn>esc</btn> was still broken.
After re-entering insert mode I could no longer leave by using <btn>esc</btn>.

I'm using IdeaVim version 1.93, Copilot 1.1.6.322 and PyCharm 2021.3.1 (#PY-213.6461.77)

When everything does work normally however, copilot still doesn't hide the results when pressing <btn>esc</btn> but it does exit insert mode. So it seems that the button is somehow not reaching copilot.

**Reply by jansorg** [reply]
@WoLpH 
I was able to reproduce the behaviour that ESC isn't removing the Copilot hints. But I'm unable to reproduce the issue where it's stuck in insert mode.

What should happen when you're in insert mode with visible Copilot hints? Should a single keypress of Esc remove the hint AND leave insert mode? Or would you expect to press Esc twice to first remove the hint and again to leave insert mode?

**Reply by wolph** [reply]
I think the preferable solution would be to have a single Esc remove the hint AND leave insert mode. If you are not in insert mode, you cannot autocomplete anyhow as far as I know so you will have to re-enter insert mode to be able to autocomplete. At least, that's how it's working for me right now.

Either solution would be fine at this point though :)

I'm wondering if it's due to some local issue with my keymaps or settings, so I've reset them to the stock keymaps for now.

**Reply by jansorg** [reply]
Thanks! I only could find a listener in the IdeaVim plugin to tell me when it entering insert mode, but nothing when it's leaving the mode. I'll think about another solution.
IdeaVim 1.9.3 changed a few things related to Esc handling, which may have broken things -- but I'm not sure.

**Reply by kurtbuilds** [reply]
Wanted to chime in on this, as this is still fairly buggy. There are a bunch of comments now that I haven't read through.

1. As far as I can tell, there's still conflict if the user has "Esc" to clear autocomplete (presumably with "Esc" for IdeaVIM). If I have IdeaVIM on, that hotkey does nothing, leaving suggestions on the screen. I can fix this by mapping the Copilot hotkey to "Cmd+Esc", but it'd be nice for it to work with a regular "Esc".

2. Currently, the plugin displays autocompletions regardless of whether the person is in insert mode. Fixing this would seemingly fix (1) at the same time.

I think this is really poor UX. I'm not trying to insert more text if I'm NOT in insert mode, so there's no reason to suggest text to insert from Copilot. Here's a particularly confusing example:

![image](https://user-images.githubusercontent.com/79804461/151278946-3550bfbb-58b5-4797-8250-816bd2815db5.png)

The block cursor shows I'm not in insert mode. The text after the cursor (`a command`) doesn't actually exist, and is a Copilot suggestion, but it's there anyway. I have to press "Cmd+Esc" to clear it, but if I wasn't paying attention and saw the text magically appear on the screen from Copilot, I might not know that the text doesn't actually exist in the document.

I think the solution is that text should only be suggested while in insert mode.

3. I'm not sure if there's been a recent change to suggestion eagerness, but recently, I find Copilot to be zealous to the point of annoyance with its suggestions. It gives suggestions extremely quickly, and it seems like it no longer matches closely enough what I write. Example here: 

![image](https://user-images.githubusercontent.com/79804461/151281300-05e61d4c-cf80-4249-981f-aae9573a51ce.png)

I saw the suggestion texts (`paths`) as I was typing, but I intentionally kept typing past it because I don't want it. For some reason, Copilot keeps recommending the same code even though I've already signaled my intent to reject it.

I can't immediately reproduce it, but I've seen Copilot give zanier recommendations (especially when I edit in between existing text) where its trying to fill in gaps, giving completely non-sensical results. I *think* I've seen it screw up the rendering as well, where IntelliJ is somehow rendering a suggestion word on top of existing text.

I think the proper behavior here is that Copilot needs to explicitly use the user text as a "prefix" and not to do partial/fuzzy matching. 

(OS X 11.6.2; IntelliJ Ultimate 2021.3; Copilot 1.1.6.322)

**Reply by jansorg** [reply]
@WoLpH cc @kurtbuilds Thank you for your input on Escape handling.
Unfortunately, the IdeaVim plugin hooks deeply into keyboard handling, which conflicts with GitHub Copilot's Escape handling. IdeaVim doesn't allow to be notified when a user exits insert mode. That means that there's currently no way to remove the hints AND to leave insert mode at the same time. I created https://youtrack.jetbrains.com/issue/VIM-2543 to tell the IdeaVim team about it.

The best we can do until this is solved in IdeaVim is to let the 1st press of Escape hide the hints and the 2nd press exit insert mode. 
I'll make that change for the next update.

**Reply by wolph** [reply]
Could we do it the other way around perhaps? Let copilot send a message to ideavim to exit insert mode?

Not ideal perhaps, but it could work around it.

Alternatively, perhaps we can use a vim autocommand to notify copilot:

`autocmd! InsertLeave * call StopCopilotCompletion`

**Reply by jansorg** [reply]
@WoLpH I already debugged the integration extensively and couldn't find a reliable way to programatically tell IdeaVim to exit insert mode. What I had made the first keypress stuck after programmatically leaving insert mode. Without official support this would always be a hack and currently it has unwanted side-effects I can't get rid of.
Can you refer to any documentation what kind of commands `autocomd` refers to? Thanks!

**Reply by wolph** [reply]
This document has all autocmd directives and documentation and how to use them: http://vimdoc.sourceforge.net/htmldoc/autocmd.html#InsertLeave

With that a line in the ideavimrc might do the trick

**Reply by jansorg** [reply]
@WoLpH Thanks! In the meantime I seem to have found a hopefully reliable way to hide hints after insert mode was exited. This change should be in tomorrow's nightly build, I'll post here when it's ready to be tested.

**Reply by wolph** [reply]
Fantastic! Thank you. I look forward to testing it :)

**Reply by kurtbuilds** [reply]
Thank you! Excited to try it out. 

So it’s not lost in the noise, any word on stopping copilot from interleaving suggestions with existing text? I recognize it might make sense to file a separate ticket. Example screenshots from myself and others are in the other thread on this page.

**Reply by jansorg** [reply]
@kurtbuilds Thanks for your other, very detailed report. It's not lost, but tracked internally as a separate issue. This isn't anything new (well, it hopefully isn't) and unfortunately to fix reliably. Interleaving with existing code on the line is useful for cases where existing code is also in the Copilot hint, but apparently it's a bit too much in a bunch of cases.

**Reply by kurtbuilds** [reply]
Got it. Happy to know it’s being worked on then! FWIW, it *seems* to me that if the rest of the line isn’t an exact match, then it makes sense to me to stop the suggestion. In my usage, if I *want* autocomplete, I’ll clear the rest of the line to see what it gives me, then undo if I want to keep it after all. The interleaving I find to be too many signals at once and too confusing. Put another way, Copilot is a value-add, but it’s lack of activating doesn’t block my coding, whereas it putting up suggestions that confuse me and slow me down reading existing code does block my coding, so I think it’s better to be mildly conservative of when it gives suggestions rather than mildly aggressive. No need to discuss further here unless it’s helpful though, it sounds like it’s already being addressed elsewhere.

Looking forward to trying out the version with only insert mode suggestions soon!

**Reply by jansorg** [reply]
@WoLpH @kurtbuilds The latest nightly at https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/nightly/155129 has the improvements for IdeaVim support. Please let us know if you're still experiencing problems with it! You can switch to the nightly channel at `Preferences > Languages & Frameworks > GitHub Copilot> Update Channel`.

**Reply by wolph** [reply]
It's definitely working far better and it properly hides the autocompletions when needed now. I'm still getting stuck in insert mode however, but I'm starting to wonder if it's not due to ideavim weirdness and/or incompatibility with my settings.

**Reply by wolph** [reply]
I can confirm that the issue is not in copilot. I was just stuck in insert mode again while I had the copilot plugin disabled. So as far as I can see all copilot specific issues have been resolved :)

Except for the interleaving, but that's not strictly a bug

**Comment by wolph** [comment]
> I think I've seen it screw up the rendering as well, where IntelliJ is somehow rendering a suggestion word on top of existing text.

I can at least say that I've seen interleaved results multiple times (i.e. 1 character real, 1 character copilot, 1 character real, 1 character copilot ... repeat for a full line).

**Reply by kurtbuilds** [reply]
Yeah, here's a great (awful) example I just encountered:

<img width="233" alt="Screen Shot 2022-01-26 at 22 12 14" src="https://user-images.githubusercontent.com/79804461/151285555-476544c7-7dbf-4ea7-bf9c-a04043385b74.png">

Not sure why the cursor isn't visible in the screenshot, but this is in normal mode (not insert mode).

**Reply by wolph** [reply]
I just got one too, in insert mode:
![image](https://user-images.githubusercontent.com/270571/151286201-2b8ca69f-8b4a-4777-9ebf-5d9df4f688d9.png)

**Comment by zablondawit** [comment]
Is this still an issue, I'm encountering this and I just switched to Copilot nightly? and also tried to remap `Hide Completions in Editor` command to `<cmd><esc>`

> This was bothering me too. You can work around it quite easily by modifying the keyboard shortcut for Github Copilot -> Hide Completions in Editor from the default value of escape. I updated it to cmd+escape, so it's still easy to dismiss the suggestions, but it no longer interferes with vim mode escape to normal mode.
> 
> <img alt="Screen Shot 2021-11-19 at 13 32 50" width="909" src="https://user-images.githubusercontent.com/79804461/142673803-671f161d-7219-4059-9b9c-280bead2cfb6.png">

**Reply by wolph** [reply]
I haven't seen the issue recently

**Reply by zachbroad** [reply]
Having this issue too

**Comment by jongoestoeleven** [comment]
The problem resurfaced, and the KeyMap entry for CoPilot plugin to Hide Completes no longer exists.
